RMarkdown: Add run & navigation commands. More customization. Refactor.

kar9222 opened the pull request:
This is my first time playing around with Typescript/javascript. So please review the codes in details thanks! 😅 😂 🤣

__What problem did you solve?__

- Add and refactor RMarkdown navigation and run commands for both extension and RMarkdown code lens.
- User-specified RMarkdown code lens commands and its orders (that is, code lens respect user-specified orders) via settings UI (RMarkdown Code Lens: Option) or settings.json (`r.rMarkdownCodeLens.option`). See below.
- More customizations
    - Enable/disable RMarkdown CodeLens, which are inline commands/buttons e.g. 'Run Chunk | Run Above' shown on the first line of each code chunk.
    - Customize chunk background color
- User-friendly options message for new users (see below)

Also closes #262

| Command|                      Keybinding|         Notes|
| ----------------------------| ------------------| --------------|
| `r.selectCurrentChunk`|       shift+alt+s| e.g. for delete/copy/cut
| `r.runCurrentChunk`|          ctrl+shift+enter| (previously implemented)
| `r.runAboveChunks`|           ctrl+alt+p|       (previously implemented)
| `r.runCurrentAndBelowChunks`| ctrl+alt+n|
| `r.runBelowChunks`|           ctrl+alt+e|
| `r.runPreviousChunk`|         ctrl+shift+alt+p|
| `r.runNextChunk`|             ctrl+shift+alt+n|
| `r.runAllChunks`|             ctrl+alt+r|
| `r.goToBeginningOfChunk`|     shift+alt+b|
| `r.goToEndOfChunk`|           shift+alt+e|
| `r.goToPreviousChunk`|        shift+alt+p|
| `r.goToNextChunk`|            shift+alt+n|
| `r.goToFirstChunk`|           shift+alt+l|
| `r.goToLastChunk`|            shift+alt+n|

Animation for both
- __Add and refactor RMarkdown navigation and run commands.__
- __User-specified RMarkdown code lens options and its orders (that is, code lens respect user-specified orders)__

(__Click to enlarge__)

![rmarkdown](https://user-images.githubusercontent.com/54139483/100451050-d9e48c00-310a-11eb-9f67-fab876e984fd.gif)


Images for __User-specified RMarkdown code lens commands and its orders (that is, code lens respect user-specified orders)__

![1212](https://user-images.githubusercontent.com/54139483/100493524-b31a6a00-318b-11eb-8073-be16d9141e70.png)
![zz1](https://user-images.githubusercontent.com/54139483/100493525-b44b9700-318b-11eb-88f4-11e2ded4d9d7.png)
![zz2](https://user-images.githubusercontent.com/54139483/100493527-b57cc400-318b-11eb-8847-f6e80bf3f4ae.png)






Customize __options and order (code lens respect user-specified orders)__ via settings UI (RMarkdown Code Lens: Option) or settings.json (`r.rMarkdownCodeLens.option`).

```json
{
  "r.rmarkdown.codeLensCommands": [
    "r.runCurrentChunk",
    "r.runAboveChunks",
    "r.runCurrentAndBelowChunks",
    "goToPreviousChunk",
    "goToNextChunk",
    "selectAllChunk",
    // more...
  ]
}
```

## More customizations

- Enable/disable RMarkdown CodeLens, which are inline commands/buttons e.g. 'Run Chunk | Run Above' shown on the first line of each code chunk.
- Customize chunk background color

User-friendly message for new users

![a3](https://user-images.githubusercontent.com/54139483/100492784-c9bcc300-3183-11eb-81d8-1ef20464dd11.png)

### Enable RMarkdown CodeLens (default)

`"r.rmarkdown.enableCodeLens": true` (default)

![a4](https://user-images.githubusercontent.com/54139483/100492836-6c754180-3184-11eb-84fe-2fba5f34663d.png)


### Disable RMarkdown CodeLens

`"r.rmarkdown.enableCodeLens": false`

- Disable only RMarkdown CodeLens, leaving other CodeLens (e.g. Git lens) enabled.
- Chunk background color is also enabled by default

![hahaz](https://user-images.githubusercontent.com/54139483/100493789-b6632500-318e-11eb-9350-b8cd586d0b84.png)





### Customize chunk background colors

`r.rmarkdown.chunkBackgroundColor: "rgba(128, 128, 128, 0.1)",`

![c1](https://user-images.githubusercontent.com/54139483/100492880-12c14700-3185-11eb-8f46-85d2a114953e.png)

`r.rmarkdown.chunkBackgroundColor: "rgba(255, 165, 0, 0.1)",`

![c2](https://user-images.githubusercontent.com/54139483/100492884-18b72800-3185-11eb-9cd2-5fe999209223.png)


Disable background color
`r.rmarkdown.chunkBackgroundColor: "",`

![c3](https://user-images.githubusercontent.com/54139483/100492892-2e2c5200-3185-11eb-8ca6-b2babaf163d3.png)


# Use cases for both  `r.runFromCurrentToBelowChunks` & `r.runBelowChunks`

1. Edit current chunk, `r.runCurrentChunk` for checking values/etc and, when done, `r.runBelowChunks` from the next chunk **(excluding the current chunk)** to all chunks below
2. After checking that the overall codes are working beautifully, edit current chunk and `r.runFromCurrentToBelowChunks` from the **current chunk** to all chunks below

![use_cases](https://user-images.githubusercontent.com/54139483/100093213-524d1200-2eab-11eb-941a-637d6d8add6e.gif)


# Design decisions & known bugs

By design, these commands work for both cases when the cursor is
- within the chunk (including both 'chunk start line' and 'chunk end line')
- outside the chunk

### Case 'within the chunk'

Works as expected

### Case cursor outside of chunk

The reason for including cases for 'outside the chunk' is to make it work for
- Navigation commands: `goToPreviousChunk`, `goToNextChunk`, `goToFirstChunk`, `goToLastChunk`, which is very convenient especially for large Rmd files with many chunks and a lot of text in between chunks.
- Less useful, but still useful for run commands (e.g. e.g. `runCurrentAndBelowChunks`) and to make the behaviour consistent with navigation commands

As a result, when the cursor is outside the chunk, the definition of the 'current chunk' and 'next chunk' become blur, that is, there are no 'correct' definitions, simply because there is no definition of 'current chunk'. To make it makes more sense, when cursor is outside the chunk, for the reason that navigation commands (e.g. `goToNextChunk`) is more useful than run commands (e.g. `runCurrentAndBelowChunks`), by design, the definition of the 'current chunk' and 'next chunk' are the same--the very next chunk. That is, e.g.

- `goToNextChunk`, `runCurrentChunk` and `runNextChunk` trigger the very next chunk
- both `runCurrentAndBelowChunks` and `runBelowChunks` trigger from the very next chunk to the last chunk
- so on...

Compared to the case where
- current chunk is the next chunk
- next chunk is the next next chunk
- then, for `goToNextChunk` will jump to next next chunk when cursor is outside the chunk, which doesn't make much sense. And it also doesn't make sense to make another command for `goToCurrentChunk` just for this trade-off.

So it's a trade-off for its design. I think the overall benefits e.g. `goToNextChunk`, `goToFirstChunk` outweigh the minor 'unexpected' behaviours when the cursor is outside the chunk, especially for large Rmd files with many chunks and a lot of text in between the chunks. Code navigation is often the most time consuming part. e.g.
- after writing a lot of text in between the chunk, with just a keybinding __shift+alt+p__ `goToPreviousChunk` often saves a lot of time

Lastly, a known bug is when the cursor is below the very last chunk,
- Commands like `goToFirstChunk`, `goToLastChunk`, `runAllChunks` work as expected (in all cases) becauses these commands doesn't require the 'current line' as argument so the cursor position doesn't affect it at all.
- But I couldn't make it work for `goToPreviousChunk`, `runPreviousChunk` and `runAboveChunks`  I thought the bugs is in internal function `getCurrentChunk` but I spent some time on it there is still some bug. I will come back and work in near future if that's okay. The bug is just weird.

-------------------------------------------------------------

Tested on Windows WSL remote extension

kar9222 commented:
For this same PR, I added new commands and manual tests in my very first post above. Please re-check. Also edited the PR title.
- `r.runAllChunks`
- `r.runPreviousChunk`
- `r.runNextChunk`

renkun-ken commented on src/extension.ts:
```diff
@@ -248,7 +248,12 @@ export async function activate(context: ExtensionContext) {
+        commands.registerCommand('r.runPreviousChunk', runPreviousChunk),
+        commands.registerCommand('r.runNextChunk', runNextChunk),
         commands.registerCommand('r.runAboveChunks', runAboveChunks),
+        commands.registerCommand('r.runFromCurrentToBelowChunks', runFromCurrentToBelowChunks),
```
Might be better named `r.runCurrentAndBelowChunks`?

kar9222 replied:
Haha definitely agree! `runFromCurrentToBelowChunks` is too long. I'll change it.

renkun-ken commented:
Thanks for implementing numerous run chunks commands. Previously, I didn't factor the code so that the CodeLens commands and extension commands do not share the algorithm of chunk detection but independently does the work.

Now that there are many versions of run chunks, I guess we should implement a function that returns an array of the info of each chunk and a function that returns the chunk id of a given all chunks detected and a document position. Then all run chunk commands and CodeLens commands could be much simplified by using the functions.

Based on this, we could also implement go-to-chunk commands such as "go to previous chunk", "go to next chunk", "go to first chunk", "go to last chunk".

kar9222 commented:
Hi @renkun-ken, thanks for everything you have done for VSCode-R & {languageserver}. Yeah I am happy and will try my best to help out a bit for the stuff that I used everyday 😄

This is actually my first time playing around with Typescript/Javascript. I have no experience with object-oriented languages too. Happy to learn if it's not too hard. The stuffs I did here were mainly based on my experience implementing some custom Vimscript for my Neovim editor e.g. finding patterns in documents, with very almost no/minimal understanding of Typescript and internals of VSCode extensions. So just a heads up that if the refactoring is too hard then I might need some time to learn a bit of Typescript from scratch?

So basically I just scanned through your Typescript implementations and playing around a bit here and there lol. So I'm actually a bit lost when you mention about __CodeLens commands__ and __extension commands__, etc. I actually don't know what is a 'codelens' lol.

Please bare with me while I'm trying to understand what you mean here. I just read through a bit of __extension.ts__ and __rmarkdown.ts__. So the _CodeLens_ are actually the _buttons_ for Rmd docs e.g. `Run Chunk` and `Run Above` implemented in `RMarkdownCodeLensProvider`.

Erm...So basically you want to refactor reusable parts

I was actually trying to figure out what you mean...in details...an hour ago but I just thought that it might be easier to try a rough refactoring and see if that's what you wanted. Because the hard part is figuring out the chunk where the cursor is positioned and also covering the cases for when the cursor is within or outside the chunk.

I just did a rough WIP commit for refactoring `runFromCurrentToBelowChunks` and it works. Do you mind having a look? NOTE the `TODO` mark in __rmarkdown.ts__

renkun-ken commented:
CodeLens commands are inline commands "Run Chunk | Run Above" showing in the document above each R chunk while extension commands are commands declared in `packages.json` and registered on extesion activation.

renkun-ken commented:
I could imagine a cleaner approach is to implement something like:

```typescript
interface RMarkdownChunk {
  id: number;
  startLine: number;
  endLine: number;
  language: string;
  options: string;
  eval: boolean;
  chunkRange: Range;
  codeRange: Range;
}

function getChunks(document: TextDocument): RMarkdownChunk[]

function getCurrentChunk(chunks: RMarkdownChunk[], line: number): RMarkdownChunk
function getPreviousChunk(chunks: RMarkdownChunk[], line: number): RMarkdownChunk
function getNextChunk(chunks: RMarkdownChunk[], line: number): RMarkdownChunk

function runCurrentChunk(chunks: RMarkdownChunk[], line: number)
function runPreviousChunk(chunks: RMarkdownChunk[], line: number)
function runNextChunk(chunks: RMarkdownChunk[], line: number)

function runAboveChunks(chunks: RMarkdownChunk[], line: number)
function runAboveAndCurrentChunks(chunks: RMarkdownChunk[], line: number)
function runBelowChunks(chunks: RMarkdownChunk[], line: number)
function runCurrentAndBelowChunks(chunks: RMarkdownChunk[], line: number)
```

where `getChunks` scans the document and returns all chunks, and all other functions just does some array filtering and mapping to obtain the code ranges of the target chunk(s) and run the code ranges in term.

kar9222 commented:
Thanks very much. That looks good.

kar9222 commented:
A quick update on the latest WIP: Refactor commit  https://github.com/Ikuyadeu/vscode-R/pull/465/commits/990bb5b3373409adcfee58ef7a522b2e1c08987f 

# Very briefly
- I got them working for both extension commands and codeLends commands.
- Added additional commands `goToPreviousChunk`, `goToNextChunk`, `goToFirstChunk`, `goToLastChunk`

# TODO
- Review the whole refactoring
- decide on the default code lens commands since now we have a lot more options
- I also provided a user-specified Rmd codeLens options `r.rmd.codeLens.option` as an array of strings (e.g. 'r.runCurrentChunk`) for more customization. But it isn't working yet. Not sure why `rmdCodeLensOpt` isn't returning values properly? Any idea? https://github.com/Ikuyadeu/vscode-R/blob/990bb5b3373409adcfee58ef7a522b2e1c08987f/src/rmarkdown.ts#L115-L117

kar9222 commented:
The latest WIP GIF for both extension commands and codeLens buttons

`r.goToPreviousChunk`: "Shift+alt+p"
`r.goToNextChunk`:     "Shift+alt+n"
`r.goToFirstChunk`:    "Shift+alt+l"
`r.goToLastChunk`:     "Shift+alt+n"
    
`r.runAboveChunks`:           "Ctrl+alt+p"
`r.runCurrentAndBelowChunks`: "Ctrl+alt+n"
`r.runBelowChunks`:           "Ctrl+alt+e"
`r.runPreviousChunk`:         "Ctrl+shift+alt+p"
`r.runNextChunk`:             "Ctrl+shift+alt+n"


![latest](https://user-images.githubusercontent.com/54139483/100371214-8bc47f80-305b-11eb-9e32-623c931b3df1.gif)

kar9222 commented:
For `goToPrevious` and other `goTo***`, I intentionally set it as 1 line below __chunk start line__ as most of the time we want to edit the codes directly instead of the chunk options? Or should I set it exactly at the __chunk start line__?

renkun-ken commented:
> For `goToPrevious` and other `goTo***`, I intentionally set it as 1 line below **chunk start line** as most of the time we want to edit the codes directly instead of the chunk options? Or should I set it exactly at the **chunk start line**?

I believe it makes good sense.

kar9222 commented:
> * I also provided a user-specified Rmd codeLens options `r.rmd.codeLens.option` as an array of strings (e.g. 'r.runCurrentChunk' specified as string in settings.json) for more customization. But it isn't working yet. Not sure why `rmdCodeLensOpt` isn't returning values properly? Any idea? https://github.com/Ikuyadeu/vscode-R/blob/990bb5b3373409adcfee58ef7a522b2e1c08987f/src/rmarkdown.ts#L115-L117

It seems like values of `r.rmd.codeLens.option` can't being accessed? Things I tried with on my machine with `import { config } from './util';`

Type __array__ default settings

```json
// Default package.json
"r.rmd.codeLens.option": {
  "type": "array",
  "default": [
    "r.runCurrentChunk",
    "r.runAboveChunks",
    "r.runCurrentAndBelowChunks",
    "r.runAllChunks",
    "r.goToPreviousChunk",
    "r.goToNextChunk"
  ],
  "description": "R RMarkdown CodeLens options.",
  "items": {
    "type": "string"
  }
}
```

```ts
import { config } from './util';

// This works
const rmdCodeLensOpt = ['r.runAllChunks', 'r.goToPreviousChunk'];

// Doesn't work
const rmdCodeLensOpt: string[] = config().get('r.rmd.codeLens.option');

return this.codeLenses.filter(e => rmdCodeLensOpt.includes(e.command.command));
```

...which produces

![1](https://user-images.githubusercontent.com/54139483/100410572-4176e900-30c3-11eb-9ca1-556e58a557e0.png)


Similarly, for type __string__ default settings

```json
// Default package.json
"r.rmd.codeLens.option": {
  "type": "string",
  "default": "r.goToNextChunk",
  "description": "test"
}
```

```ts
// This work
const rmdCodeLensOpt = 'r.runAllChunks';

// Doesn't work
const rmdCodeLensOpt: string = config().get('r.rmd.codeLens.option');

return this.codeLenses.filter(e => e.command.command === rmdCodeLensOpt);
```

kar9222 commented:
Oh nvm I fixed it after reading docs of `workspace.getConfiguration` in VSCode module. It should be `rmd.codeLens.option` instead of `r.rmd.codeLens.option`

kar9222 commented:
Personally I turn off codelens as I mostly use keyboard. But now users who use buttons can now easily customize their preferred RMarkdown codelens buttons to show directly in their editor e.g. they can use keybindings for commonly used commands like `runCurrentChunk` while use buttons for occasionally used commands like `runCurrentAndBelowChunks` or `runAllChunks`. It's best of both world for them. I list all available options in the settings UI and they can even customize on the fly lol. See GIF below.

I still can't believe VSCode is so customizable. The API is superb too.

This is the package.json options

```json
{
  "r.rMarkdownCodeLens.option": {
    "type": "array",
    "items": {
      "type": "string"
    },
    "default": [
      "r.runCurrentChunk",
      "r.runAboveChunks",
      "r.runCurrentAndBelowChunks",
      "r.runAllChunks",
      "r.goToPreviousChunk",
      "r.goToNextChunk"
    ],
    "description": "Available options:\n\nr.runCurrentChunk\nr.runAboveChunks\nr.runCurrentAndBelowChunks\nr.runBelowChunks\nr.runAllChunks\nr.runPreviousChunk\nr.runNextChunk\nr.goToPreviousChunk\nr.goToNextChunk\nr.goToFirstChunk\nr.goToLastChunk\n\n----------------------------------\nSpecify options:"
  },

}
```

![rmd_codelens_options](https://user-images.githubusercontent.com/54139483/100414553-18a82100-30ce-11eb-9593-704d0c584472.gif)

kar9222 commented:
Added 
- `goToBeginningOfChunk`
- `goToEndOfChunk`
- `selectCurrentChunk` (e.g. for copy, delete, cut)


![new_commands_2](https://user-images.githubusercontent.com/54139483/100426597-4056b380-30e5-11eb-94d0-8af84d82dff0.gif)

kar9222 commented:
Implemented user-specified sort order for RMarkdown code lenses.

User-friendly message for customizing both the options and its orders
![103](https://user-images.githubusercontent.com/54139483/100442231-c847b800-30fb-11eb-802b-496512352ff4.png)

![100](https://user-images.githubusercontent.com/54139483/100442241-ca117b80-30fb-11eb-85c9-07bc46a5b5c8.png)

![101](https://user-images.githubusercontent.com/54139483/100442363-f9c08380-30fb-11eb-889a-5b1821b0f63d.png)

kar9222 commented:
Updated my first post above, including summary table of commands, latest GIF and images...

kar9222 commented:
## Add options: enable/disable CodeLens, chunk background color

More customization e.g. personally I disable CodeLens and chunk background color for minimalistic view. Users may want to customize chunk background color e.g. lighter/darker color.

### Default (no settings) 

![true](https://user-images.githubusercontent.com/54139483/100491206-dafed300-3175-11eb-9fd4-1be10a5906bc.png)

### Disable CodeLens globally `"[rmd]": { "editor.codeLens": false }`

Workaround as mentioned in https://github.com/Ikuyadeu/vscode-R/pull/429#issuecomment-718278133, it disables all CodeLens including RMarkdown CodeLens and GitLens.

![false](https://user-images.githubusercontent.com/54139483/100491210-ec47df80-3175-11eb-8590-f356c1c695d6.png)

## New options: enable/disable Rmd CodeLens, chunk background color

User-friendly message for new users

![4455](https://user-images.githubusercontent.com/54139483/100491231-14374300-3176-11eb-9bce-e7a826d96a60.png)
 
### `"r.rMarkdownCodeLens.enableCodeLens": true` (default)

Enable RMarkdown CodeLens

![333](https://user-images.githubusercontent.com/54139483/100491246-2f09b780-3176-11eb-83e9-2127fffab843.png)

### `"r.rMarkdownCodeLens.enableCodeLens": false`

Disable RMarkdown CodeLens

![444](https://user-images.githubusercontent.com/54139483/100491247-3204a800-3176-11eb-962d-80353c2e0d94.png)

### Chunk background colors

![t555](https://user-images.githubusercontent.com/54139483/100491442-907e5600-3177-11eb-819e-bfe8f53e3ca2.png)

![t666](https://user-images.githubusercontent.com/54139483/100491452-aab83400-3177-11eb-9b6a-928878b1fb55.png)

renkun-ken commented on package.json:
```diff
@@ -511,6 +655,31 @@
             "type": "string"
           }
         },
+        "r.rMarkdownCodeLens.option": {
```
Does it make sense that the settings for RMarkdown is named under "r.rmarkdown.*" as `rmarkdown` is often considered a term in this case?

renkun-ken replied:
Also, this particular setting is about the chunk commands provided by CodeLens, I suggest that it is named `r.rmarkdown.codeLensCommands`? Or alternatively, `r.rmarkdown.chunkCommands` as many users might not know what is CodeLens?

kar9222 replied:
Yea totally agree. Renamed to `r.rmarkdown.codeLensOptions`

kar9222 replied:
This isn't resolved yet. I think `r.rmarkdown.codeLensCommands` sounds better than `r.rmarkdown.chunkCommands` because `chunkCommands` is more generic, but the settings is specific to CodeLens. What do you think?

kar9222 replied:
`r.rmarkdown.enableCodeLens` also refers to CodeLens, so `r.rmarkdown.codeLensCommands` sounds more consistent with it.

kar9222 replied:
I added some description for this

![99](https://user-images.githubusercontent.com/54139483/100493420-c37e1500-318a-11eb-8553-86d9cae71fa6.png)

renkun-ken replied:
Yes, I agree. `r.rmarkdown.enableCodeLens` and `r.rmarkdown.codeLensCommands` sound more consistent.

renkun-ken commented on package.json:
```diff
@@ -511,6 +655,31 @@
+          ],
+          "description": "Available options:\n\nr.selectCurrentChunk\nr.runCurrentChunk\nr.runAboveChunks\nr.runCurrentAndBelowChunks\nr.runBelowChunks\nr.runAllChunks\nr.runPreviousChunk\nr.runNextChunk\nr.goToBeginningOfChunk\nr.goToEndOfChunk\nr.goToPreviousChunk\nr.goToNextChunk\nr.goToFirstChunk\nr.goToLastChunk\n\n----------------------------------\nCustomize both the options AND its orders ...
+        },
+        "r.rMarkdownCodeLens.enableCodeLens": {
```
Suggest `r.rmarkdown.enableCodeLens`.

kar9222 replied:
Renamed to `r.rmarkdown.enableCodeLens`

renkun-ken commented on package.json:
```diff
@@ -511,6 +655,31 @@
+          "default": true,
+          "description": "Enable RMarkdown code lens"
+        },
+        "r.rMarkdownCodeLens.chunkBackgroundColor": {
```
Here it seems to make more sense to use `r.rmarkdown.chunkBackgroundColor` in this case as the background color of chunks is handled by document decoration rather than being part of CodeLens, i.e. it has nothing to do with CodeLens but it part of settings of R Markdown document rendering.

kar9222 replied:
Thanks for the review. Renamed to `r.rmarkdown.chunkBackgroundColor` and made the customization of CodeLens and chunk background color independent

renkun-ken commented on package.json:
```diff
@@ -511,6 +655,31 @@
+        "r.rMarkdownCodeLens.chunkBackgroundColor": {
+          "type": "string",
+          "default": "rgba(128, 128, 128, 0.1)",
+          "description": "RMarkdown code lens chunk background color in RGBA or RGB value. Defaults to rgba(128, 128, 128, 0.1). Reload VS Code after changing settings. Only works when option `rMarkdownCodeLens.enableCodeLens` is set to true (default).\n\nLearn how to set colors: https://www.w3schools.com/css/css_colors_rgb.asp.\n\nExamples for syntax rgba(<red>, <green>, <blue>, <alpha>):\nrgba( ...
```
See https://github.com/Ikuyadeu/vscode-R/pull/465/files#r531823145. CodeLens and background decoration are separate things and are independent of each other.

kar9222 replied:
Made the customization of CodeLens and chunk background color independent

renkun-ken commented on package.json:
```diff
@@ -511,6 +655,31 @@
+        "r.rMarkdownCodeLens.enableCodeLens": {
+          "type": "boolean",
+          "default": true,
+          "description": "Enable RMarkdown code lens"
```
Might need to say more about it since many users don't know what is CodeLens.

kar9222 replied:
Added more detailed description as per the latest commit.

kar9222 commented:
Kun Thanks for the review. I pushed the changes in the latest commit. For manual tests, I also edited all related-images in the first post.

- Renamed options in package.json
- Edited descriptions in package.json
- Allow users to customize CodeLens & chunk color __independently__ e.g. disable RMarkdown CodeLens but enable chunk background color (default) by simply setting `"r.rmarkdown.enableCodeLens": false`. This solves the previous commit that disabling Codelens also disable chunk background color. cc @MilesMcBain who wants this customization

kar9222 commented:
I added quite verbose description for the settings because I don't think new users know the jargon `CodeLens`. I myself didn't even know about it either until I started this PR 😅

kar9222 commented:
There is no more new features I want to add. So it's pending review and ready to be merged.

Things to be reviewed:

#### 1. Code refactoring in general...

#### 2. Default CodeLens commands and keybindings
Currently, I set these because I think setting more commands as default is a good way for new users to explore and play around. Otherwise, they might not even discover many of these useful commands, given that VSCode has so many keybindings. I wanted to add `selectCurrentChunk` but it seems a bit too much. So I just left it out. They can always customize it anyway.

| Command|                      Keybinding
| ----------------------------| ------------------
| `r.runCurrentChunk`|          ctrl+shift+enter
| `r.runAboveChunks`|           ctrl+alt+p
| `r.runCurrentAndBelowChunks`| ctrl+alt+n
| `r.runAllChunks`|             ctrl+alt+r
| `r.goToPreviousChunk`|        shift+alt+p
| `r.goToNextChunk`|            shift+alt+n

renkun-ken commented:
Thanks for the nice work!

I'll review and test later today, mostly to find out bugs.

kar9222 commented:
By design, these commands work for both cases when the cursor is
- within the chunk (including both 'chunk start line' and 'chunk end line')
- outside the chunk

Case 'within the chunk' works as expected.

The reason for including cases for 'outside the chunk' is to make it work for
- Navigation commands: `goToPreviousChunk`, `goToNextChunk`, `goToFirstChunk`, `goToLastChunk`, which is very convenient especially for large Rmd files.
- Less useful, but still useful for run commands (e.g. e.g. `runCurrentAndBelowChunks`) and to make the behaviour consistent with navigation commands

As a result, when the cursor is outside the chunk, the definition of the 'current chunk' and 'next chunk' become blur, that is, there are no 'correct' definitions, simply because there is no 'current chunk'. To make it makes more sense, when cursor is outside the chunk, for the reason that navigation commands (e.g. `goToNextChunk`) is more useful than run commands (e.g. `runCurrentAndBelowChunks`), by design, the definition of the 'current chunk' and 'next chunk' are the same--the very next chunk. That is, e.g.

- `goToNextChunk`, `runCurrentChunk` and `runNextChunk` trigger the very next chunk
- both `runCurrentAndBelowChunks` and `runBelowChunks` trigger from the very next chunk to the last chunk
- so on...

Compared to the case where
- current chunk is the next chunk
- next chunk is the next next chunk
- then, for `goToNextChunk` will jump to next next chunk when cursor is outside the chunk, which doesn't make much sense. And it also doesn't make sense to make another command for `goToCurrentChunk` just for this trade-off.

So it's a trade-off for its design. I think the overall benefits e.g. `goToNextChunk`, `goToFirstChunk` outweigh the minor 'unexpected' behaviours when the cursor is outside the chunk, especially for large Rmd files with many chunks and a lot of text in between the chunks. Code navigation is often the most time consuming part. e.g.
- after writing a lot of text in between the chunk, with just a keybinding __shift+alt+p__ `goToPreviousChunk` often saves a lot of time

Lastly, a known bug is when the cursor is below the very last chunk,
- Commands like `goToFirstChunk`, `goToLastChunk`, `runAllChunks` work as expected (in all cases) becauses these commands doesn't require the 'current line' as argument so the cursor position doesn't affect it at all.
- But I couldn't make it work for `goToPreviousChunk`, `runPreviousChunk` and `runAboveChunks`  I thought the bugs is in `getCurrentChunk` but I spent some time on it there is still some bug. I will come back and work in near future if that's okay. The bug is just weird.

kar9222 commented:
I documented the design decision and bugs mentioned above in the 1st post. Updated.

Ikuyadeu commented:
@kar9222 Thank you for your hard works.

I'm sorry that you did a lot of work, but could you remove some features?
Too many commands and shortcuts can interfere with other vscode-extensions commands and cause confusion for beginners.
At least I think there can be too many features that don't exist in R Studio, vscode-jupyter, vscode-python.

1.
commands
* goToFirstChunk
* goToLastChunk
* goToBeginningOfChunk
* goToEndOfChunk

2. The default `codeLensCommand` option are
             "r.runCurrentChunk",
             "r.runAboveChunks"
Delete other than the above

3. All key bindings can be deleted, customized keys can be used by writing them on the wiki, or by creating a keybindings plugin.

kar9222 commented:
Hi @Ikuyadeu , thank you very much for the review and this extensions. I'm happy to help out a little bit for the stuffs that I use everyday 👍

As per your request, I
- removed `goToFirstChunk`, `goToLastChunk`, `goToBeginningOfChunk`, `goToEndOfChunk`
- default CodeLens to `runCurrentChunk`, `runAboveChunks`
- removed all keybindings except for `runCurrentChunk`, `runAboveChunks`

Let me know if anything needs change. Thank you.

Ikuyadeu commented on src/rmarkdown.ts:
```diff
@@ -4,6 +4,8 @@ import {
   Event, EventEmitter, Position, Range, TextDocument, TextEditorDecorationType, window
 } from 'vscode';
 import { runChunksInTerm } from './rTerminal';
+import * as vscode from 'vscode';
```
Can you import `Selection` by importing from line 4?

kar9222 replied:
Yes. Sorry for mistake. Done in https://github.com/Ikuyadeu/vscode-R/pull/465/commits/1d418d8a3767c2367702933089239048c059ad7e

Ikuyadeu commented on src/rmarkdown.ts:
```diff
@@ -112,95 +139,275 @@ export class RMarkdownCodeLensProvider implements CodeLensProvider {
+function getCurrentChunk__CursorWithinChunk(chunks: RMarkdownChunk[], line: number): RMarkdownChunk {
+  let id = 0;
+
+  while (id <= chunks.length - 1) {
```
It can use for-of statemente

kar9222 replied:
Done in https://github.com/Ikuyadeu/vscode-R/pull/465/commits/213bad2e090c4ae470f019e335b0eedb491c257d

Ikuyadeu commented on src/rmarkdown.ts:
```diff
@@ -43,67 +44,93 @@ export class RMarkdownCodeLensProvider implements CodeLensProvider {
+      chunkRanges.push(chunkRange);
+
+      // Enable/disable only CodeLens, without affecting chunk background color.
+      if (config().get<boolean>('rmarkdown.enableCodeLens')) {
```
here can use && symbol

kar9222 replied:
Done in https://github.com/Ikuyadeu/vscode-R/pull/465/commits/5309e8c60dfa73992a1e6c36209e10df5b7a1e47

Ikuyadeu commented:
Thank you for your early fix.
LGTM.

renkun-ken commented:
For the following rmd:

````
# Title

This is a test document.

## Section 1

```{r}
1
```

```{r}
2
```

```{r,eval=FALSE}
3
```

```{python}
def fun(x,y):
  return x+y
```

```{sql}
select * from table
```

```{R chunk-1}
4
```

## Section 2

```{r plot1}
5
```

```{r plot2}
6
```
````

The run chunks commands seem to ignore chunk language, i.e. they run chunks of all languages. Instead, they should only send R chunks to the terminal.

Also, some edge cases are not properly handled yet. For example, Run Above the first chunk, or Run Below the last chunk will cause errors like

<img width="468" alt="image" src="https://user-images.githubusercontent.com/4662568/100516536-9009ad80-31bf-11eb-8932-a2c0393d1a2e.png">

Ikuyadeu commented:
@renkun-ken Thank you for your test, I made other issues.
Your posted issue #473 
My refactoring issue #472

kar9222 commented:
@renkun-ken 

- `selectCurrentChunk` by design, it internally uses `getCurrentChunk__CursorWithinChunk` so that it only works when cursor is within the chunk the command doesn't make sense when cursor is outside the chunk. So error handling is the best we can do? What is the typescript way of handling error?
- When cursor is at the first (last) chunk, `runAboveChunks` (`runCurrentAndBelowChunks` and `runBelowChunks`) should have error handling.
- When cursor is below the last chunk, it's a known bug mentioned in https://github.com/Ikuyadeu/vscode-R/pull/465#issuecomment-735101521
- for non-R languages, I'm not sure why it doesn't work and I thought it's handled by https://github.com/Ikuyadeu/vscode-R/blob/3e5822b0e3bf1025e77a79aaf90e72af8db948c7/src/rmarkdown.ts#L65

kar9222 commented:
@renkun-ken I figured out the non-R languages bugs, this line https://github.com/Ikuyadeu/vscode-R/blob/3e5822b0e3bf1025e77a79aaf90e72af8db948c7/src/rmarkdown.ts#L65

only handles the CodeLens but doesn't handle the extension commands, which can be hot fixed by e.g.

```ts
// Original
export async function runCurrentChunk(chunks: RMarkdownChunk[] = _getChunks(),
                                      line: number = _getStartLine()) {
  const currentChunk = getCurrentChunk(chunks, line);
    runChunksInTerm([currentChunk.codeRange]);
}

export async function runCurrentChunk(chunks: RMarkdownChunk[] = _getChunks(),
                                      line: number = _getStartLine()) {
  const currentChunk = getCurrentChunk(chunks, line);
  if (currentChunk.language === 'r') {
    return runChunksInTerm([currentChunk.codeRange]);
  }
}
```

kar9222 commented:
If you think CodeLens and extension commands should be separated, then we can make a helper e.g. 

```ts
function runChunksInTerm__R(chunk: RMarkdownChunk, range: Range) {
  if (chunk.language === 'r') {
    return runChunksInTerm([range]);
  }
}
```
